core/src/main/java/io/questdb/cutlass/http/HttpSessionStoreImpl.java

```diff
@@ -91,24 +125,24 @@ public synchronized int size(@NotNull CharSequence principal) {
 
     @Override
     public void updateUserGroups(@NotNull CharSequence principal, @NotNull ObjList<CharSequence> groups) {
-        final ObjList<CharSequence> currentGroups = groupsByEntity.get(principal);
+        final ReadOnlyObjList<CharSequence> currentGroups = groupsByEntity.get(principal);
         if (currentGroups == null) {
-            groupsByEntity.put(principal, groups);
+            groupsByEntity.put(principal, groups.copy());
             return;
         }
 
         // ideally these would be compared as sets, but it is ok
         // unlikely that the order of groups changing constantly
         if (!groups.equals(currentGroups)) {
```

Comment:
under contention, updates will get lost silently if map changes between "get" and "replace"

Reply:
This will be racy even if the entire method is made `synchronized`.
If the groups are updated in the OAuth2 provider between two HTTP calls, and one thread gets the old, and the other one gets the new value, there is no guarantee that we end up with the new value in the map.
Only guaranteed solution would be to make calls to the provider and updates to the map fully synchronized, but this is not necessary.
If the update is missed because of a race, we can still guarantee that the change will make it into the cache at the next update, after oidc client cache expiry time, and that is ok.
It is basically the same situation as if we only had one thread which has just missed the update in the provider.